File: beam-android/core/model/src/main/kotlin/dev/beam/android/core/model/LoadState.kt
package dev.beam.android.core.model

/**
 * The state of something being loaded, as a screen renders it.
 *
 * Modelled as a sealed hierarchy rather than a struct of nullable fields so a
 * screen cannot render "loading" and "error" at once, and so `when` is
 * exhaustive over the cases that actually exist.
 */
public sealed interface LoadState<out T> {

    /** Nothing has been requested yet. */
    public data object Idle : LoadState<Nothing>

    /**
     * A request is in flight.
     *
     * @property previous the last successful value, kept so a refresh can keep
     *   showing content instead of flashing a spinner over it.
     */
    public data class Loading<out T>(val previous: T? = null) : LoadState<T>

    /** The request succeeded. */
    public data class Success<out T>(val value: T) : LoadState<T>

    /**
     * The request failed.
     *
     * @property message text already phrased for a person.
     * @property retryable whether offering a retry would be honest.
     * @property previous the last successful value, so a failed refresh can
     *   keep showing stale content with an error alongside it.
     */
    public data class Failure<out T>(
        val message: String,
        val retryable: Boolean,
        val previous: T? = null,
    ) : LoadState<T>
}

/** The value held, if this state has one -- current or stale. */
public val <T> LoadState<T>.valueOrNull: T?
    get() = when (this) {
        is LoadState.Success -> value
        is LoadState.Loading -> previous
        is LoadState.Failure -> previous
        LoadState.Idle -> null
    }

/** Whether a progress indicator belongs on screen. */
public val LoadState<*>.isLoading: Boolean
    get() = this is LoadState.Loading

File: beam-apple/Packages/BeamKit/Sources/BeamModel/LoadState.swift
import Foundation

/// The lifecycle of one asynchronously loaded value.
///
/// Modelled as an enum rather than as parallel `isLoading`/`value`/`error`
/// properties, because those admit states that cannot happen -- loading and
/// failed at once, a value beside an error -- and every view then has to
/// decide which to believe. Mirrors `LoadState` in `beam-android`'s
/// `core/model`.
public enum LoadState<Value: Sendable>: Sendable {
    /// Nothing has been asked for yet.
    case idle
    /// A request is in flight and nothing has arrived.
    case loading
    /// A value arrived.
    case loaded(Value)
    /// The request failed, with a message fit to show.
    case failed(String)

    /// The loaded value, if there is one.
    public var value: Value? {
        if case .loaded(let value) = self { return value }
        return nil
    }

    /// Whether a request is in flight.
    public var isLoading: Bool {
        if case .loading = self { return true }
        return false
    }

    /// The failure message, if the request failed.
    public var failure: String? {
        if case .failed(let message) = self { return message }
        return nil
    }
}

extension LoadState: Equatable where Value: Equatable {}
